### PR TITLE
Fix MongoDb operations to use transaction on both read and write

### LINKED ISSUE #345: Remove old tag subscribers code
https://github.com/akkadotnet/Akka.Persistence.MongoDB/blob/ef7bed77e9628a6000109cb0a0f25964fa7f095f/src/Akka.Persistence.MongoDb/Journal/MongoDbJournal.cs#L299-L308

### LINKED ISSUE #346: Remove old persistence id subscribers code
https://github.com/akkadotnet/Akka.Persistence.MongoDB/blob/ef7bed77e9628a6000109cb0a0f25964fa7f095f/src/Akka.Persistence.MongoDb/Journal/MongoDbJournal.cs#L313-L314

### PR DESCRIPTION
**Pull request by Arkatufus**
Fixes #346 
Fixes #345

## Changes

* Journal
  * Remove subscription to new events, tags, etc
  * All operations are covered using a single cancellation token
  * Transactions are applied to the whole operation
  * Extract queries into its own static class
* Snapshot store
  * All operations are covered using a single cancellation token
  * Transactions are applied to the whole operation

## Checklist

For significant changes, please ensure that the following have been completed (delete if not relevant):

* [x] This change follows the [Akka.NET API Compatibility Guidelines](https://getakka.net/community/contributing/api-changes-compatibility.html).
* [ ] I have [reviewed my own pull request](https://getakka.net/community/contributing/index.html#review-your-own-pull-requests).

### Latest `dev` Benchmarks 

Include data from the [relevant benchmark](https://getakka.net/community/contributing/index.html#improve-performance) prior to this change here.

### This PR's Benchmarks

Include data from after this change here.

### CODE REVIEW

File: src/Akka.Persistence.MongoDb/Journal/MongoDbJournal.cs [hunk 1]
```diff
@@ -28,19 +29,14 @@ namespace Akka.Persistence.MongoDb.Journal
...
-        private Lazy<IMongoCollection<JournalEntry>> _journalCollection;
-        private Lazy<IMongoCollection<MetadataEntry>> _metadataCollection;
-
-        private ImmutableDictionary<string, IImmutableSet<IActorRef>> _persistenceIdSubscribers =
-            ImmutableDictionary.Create<string, IImmutableSet<IActorRef>>();
-
-        private ImmutableDictionary<string, IImmutableSet<IActorRef>> _tagSubscribers =
-            ImmutableDictionary.Create<string, IImmutableSet<IActorRef>>();
-
-        private readonly HashSet<IActorRef> _newEventsSubscriber = new();
```
Comment (Arkatufus) [inline src/Akka.Persistence.MongoDb/Journal/MongoDbJournal.cs:43]: Remove all copy-paste byte rot from sqlite era

File: src/Akka.Persistence.MongoDb/Journal/MongoDbJournal.cs [hunk 2]
```diff
@@ -28,19 +29,14 @@ namespace Akka.Persistence.MongoDb.Journal
...
+        // ReSharper disable InconsistentNaming
+        private IMongoDatabase? _mongoDatabase_DoNotUseDirectly;
+        private IMongoCollection<JournalEntry>? _journalCollection_DoNotUseDirectly;
+        private IMongoCollection<MetadataEntry>? _metadataCollection_DoNotUseDirectly;
```
Comment (Arkatufus) [inline src/Akka.Persistence.MongoDb/Journal/MongoDbJournal.cs:38]: Convert `Lazy<T>` to async getter methods

File: src/Akka.Persistence.MongoDb/Journal/MongoDbJournal.cs [hunk 3]
```diff
@@ -64,131 +60,152 @@ private MongoDbJournal(MongoDbJournalSettings settings)
             _serialization = Context.System.Serialization;
         }
 
-        protected override void PreStart()
+        private IMongoDatabase GetMongoDb()
```
Comment (Arkatufus) [inline src/Akka.Persistence.MongoDb/Journal/MongoDbJournal.cs:63]: Getter method for `IMongoDatabase`

File: src/Akka.Persistence.MongoDb/Journal/MongoDbJournal.cs [hunk 4]
```diff
@@ -64,131 +60,152 @@ private MongoDbJournal(MongoDbJournalSettings settings)
...
+            var connectionString = new MongoUrl(_settings.ConnectionString);
+            var clientSettings = MongoClientSettings.FromUrl(connectionString);
+            clientSettings.LinqProvider = LinqProvider.V2;
+            
+            client = new MongoClient(clientSettings);
+            _mongoDatabase_DoNotUseDirectly = client.GetDatabase(connectionString.DatabaseName);
+            return _mongoDatabase_DoNotUseDirectly;
+        }
+        
+        private async Task<IMongoCollection<JournalEntry>> GetJournalCollection(CancellationToken token)
```
Comment (Arkatufus) [inline src/Akka.Persistence.MongoDb/Journal/MongoDbJournal.cs:87]: Async getter method for `IMongoCollection<JournalEntry>>`

File: src/Akka.Persistence.MongoDb/Journal/MongoDbJournal.cs [hunk 5]
```diff
@@ -64,131 +60,152 @@ private MongoDbJournal(MongoDbJournalSettings settings)
...
+                    .Ascending(entry => entry.Tags)
+                    .Ascending(entry => entry.Ordering));
+
+            await _journalCollection_DoNotUseDirectly.Indexes
+                .CreateOneAsync(tagsWithOrdering, cancellationToken:token);
+
+            return _journalCollection_DoNotUseDirectly;
         }
 
+        private async Task<IMongoCollection<MetadataEntry>> GetMetadataCollection(CancellationToken token)
```
Comment (Arkatufus) [inline src/Akka.Persistence.MongoDb/Journal/MongoDbJournal.cs:128]: Async getter method for `IMongoCollection<MetadataEntry>>`

File: src/Akka.Persistence.MongoDb/Journal/MongoDbJournal.cs [hunk 6]
```diff
@@ -64,131 +60,152 @@ private MongoDbJournal(MongoDbJournalSettings settings)
...
+            }
+            
+            using var session = await GetMongoDb().Client.StartSessionAsync(EmptySessionOptions, token);
+            await session.WithTransactionAsync(
+                async (s, ct) =>
+                {
+                    await act(s, ct);
+                    return Task.FromResult(NotUsed.Instance);
+                }, cancellationToken:token);
+        }
```
Comment (Arkatufus) [inline src/Akka.Persistence.MongoDb/Journal/MongoDbJournal.cs:172]: This method wraps a persistence operation (multiple steps) inside the MongoDb built-in transaction handler

File: src/Akka.Persistence.MongoDb/Journal/MongoDbJournal.cs [hunk 7]
```diff
@@ -64,131 +60,152 @@ private MongoDbJournal(MongoDbJournalSettings settings)
...
-                    .Limit(limitValue)
-                    .ToListAsync(unitedCts.Token);
+            var journalCollection = await GetJournalCollection(unitedCts.Token);
 
-                collections.ForEach(doc => { recoveryCallback(ToPersistenceRepresentation(doc, context.Sender)); });
-            }
+            await MaybeWithTransaction(async (session, ct) =>
+            {
+                var collections = await journalCollection
+                    .ReplayMessagesQuery(session, persistenceId, fromSequenceNr, toSequenceNr, limitValue)
```
Comment (Arkatufus) [inline src/Akka.Persistence.MongoDb/Journal/MongoDbJournal.cs:204]: Some of these queries are refactored into their own static method to preserve logic consistency across both transactioned and non-transactioned calls

File: src/Akka.Persistence.MongoDb/Journal/MongoDbJournal.cs [hunk 8]
```diff
@@ -198,63 +215,43 @@ private CancellationTokenSource CreatePerCallCts()
...
-            var builder = Builders<JournalEntry>.Filter;
-            var seqNoFilter = builder.AnyEq(x => x.Tags, tag);
-            if (fromSequenceNr > 0)
-                seqNoFilter &= builder.Gt(x => x.Ordering, new BsonTimestamp(fromSequenceNr));
-            if (toSequenceNr != long.MaxValue)
-                seqNoFilter &= builder.Lte(x => x.Ordering, new BsonTimestamp(toSequenceNr));
+            using var unitedCts = CreatePerCallCts();
+            var journalCollection = await GetJournalCollection(unitedCts.Token);
 
+            return await MaybeWithTransaction(async (s, ct) =>
```
Comment (Arkatufus) [inline src/Akka.Persistence.MongoDb/Journal/MongoDbJournal.cs:221]: Transactions are also applied to the read operations and spans for the whole persistence operation

File: src/Akka.Persistence.MongoDb/Journal/MongoDbJournalQueries.cs [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using MongoDB.Bson;
+using MongoDB.Driver;
+using MongoDB.Driver.Linq;
+
+#nullable enable
+namespace Akka.Persistence.MongoDb.Journal;
+
+internal static class MongoDbJournalQueries
```
Comment (Arkatufus) [inline src/Akka.Persistence.MongoDb/Journal/MongoDbJournalQueries.cs:18]: Queries are refactored into this static class

File: src/Akka.Persistence.MongoDb/Query/AllEventsPublisher.cs [hunk 10]
```diff
@@ -114,7 +113,8 @@ protected bool Replaying(object message)
                         persistenceId: replayed.Persistent.PersistenceId,
                         sequenceNr: replayed.Persistent.SequenceNr,
                         timestamp: replayed.Persistent.Timestamp,
-                        @event: replayed.Persistent.Payload));
+                        @event: replayed.Persistent.Payload,
+                        tags: Array.Empty<string>()));
```
Comment (Arkatufus) [inline src/Akka.Persistence.MongoDb/Query/AllEventsPublisher.cs:117]: Add tag in EventEnvelope feature support

**Review by Arkatufus — COMMENTED** [review]
Self review

File: src/Akka.Persistence.MongoDb/Journal/MongoDbJournal.cs (hunk 1 above)
Comment (Aaronontheweb) [reply src/Akka.Persistence.MongoDb/Journal/MongoDbJournal.cs:43]: Ah, does this issue also close https://github.com/akkadotnet/Akka.Persistence.MongoDB/issues/346 and https://github.com/akkadotnet/Akka.Persistence.MongoDB/issues/345 ?

File: src/Akka.Persistence.MongoDb/Journal/MongoDbJournal.cs (hunk 4 above)
Comment (Aaronontheweb) [reply src/Akka.Persistence.MongoDb/Journal/MongoDbJournal.cs:87]: LGTM

File: src/Akka.Persistence.MongoDb/Journal/MongoDbJournal.cs (hunk 6 above)
Comment (Aaronontheweb) [reply src/Akka.Persistence.MongoDb/Journal/MongoDbJournal.cs:172]: LGTM

File: src/Akka.Persistence.MongoDb/Journal/MongoDbJournalQueries.cs (hunk 9 above)
Comment (Aaronontheweb) [reply src/Akka.Persistence.MongoDb/Journal/MongoDbJournalQueries.cs:18]: LGTM

File: src/Akka.Persistence.MongoDb/Query/QueryApi.cs [hunk 11]
```diff
@@ -25,6 +23,7 @@ public interface ISubscriptionCommand { }
     /// the subscriber when <see cref="AsyncWriteJournal.WriteMessagesAsync"/> has been called.
     /// </summary>
     [Serializable]
+    [Obsolete("Query is not implemented.", true)]
     public sealed class SubscribePersistenceId : ISubscriptionCommand
```
Comment (Aaronontheweb) [inline src/Akka.Persistence.MongoDb/Query/QueryApi.cs:27]: Database-driven versions of these queries, which work properly, still exist though right?

**Review by Aaronontheweb — APPROVED** [review]
LGTM but left some questions

File: src/Akka.Persistence.MongoDb/Query/QueryApi.cs (hunk 11 above)
Comment (Arkatufus) [reply src/Akka.Persistence.MongoDb/Query/QueryApi.cs:27]: That is part of `Persistence.Query` which is different than these

Comment (Arkatufus) [reply src/Akka.Persistence.MongoDb/Query/QueryApi.cs:27]: Yes, this PR would close #346 and #345, its a simple fix, we just haven't done it yet

**Comment by Aaronontheweb** [comment]
@Arkatufus FYI, tests are not running for the majority of the suite

**Comment by Arkatufus** [comment]
it should be fixed now

**Comment by Aaronontheweb** [comment]
> Akka.Persistence.MongoDb.Tests.MongoDbTransactionCurrentAllEventsSpec.ReadJournal_query_CurrentAllEvents_should_find_existing_events
Failed: Timeout 00:00:30 while waiting for a message of type System.String 

Gonna see if that fails again upon a re-run

**Comment by Arkatufus** [comment]
## Note for future developers:
As of this PR, the session object used inside a transaction is not thread safe. 
Never run 2 or more concurrent DB operations inside a transaction, they will cause the transaction to fail horribly. 
This is easily spotted by looking for any detached `Task` (any `Task` that is not `await`-ed), or the use of `Task.WhenAll` or `Task.WhenAny`.
Just. Don't.

**Pull request merged**